Add hash-mode 3730 - Dahua NVR/DVR - md5($salt1.strtoupper(md5($salt2…

davidrozen76 opened the pull request:
….$pass)))

Pure kernel on attack mode 0 doesn't crack when only a single hash is present in the hashfile (otherwise ok) - self-test fails as well.

In accordance to issue #3468

Commit:
Add hash-mode 3730 - Dahua NVR/DVR - md5($salt1.strtoupper(md5($salt2.$pass)))

Pure kernel on attack mode 0 doesn't crack when only a single hash is present in the hashfile (otherwise ok) - self-test fails as well.

piwvvo commented on OpenCL/m03730_a0-pure.cl:
```diff
@@ -0,0 +1,309 @@
+
+  md5_init (&ctx0);
+
+  md5_update_global (&ctx0, esalt_bufs[DIGESTS_OFFSET_HOST].salt1_buf, esalt_bufs[DIGESTS_OFFSET_HOST].salt1_len);
```
It seems that single hash kernel doesn't work correctly because you used salt1_buf here instead of salt2_buf

Commit:
Fixed pure kernel on attack mode 0 for hash 3730

Concatenated $salt1 instead of $salt2 to $pass, rendering single hashes uncrackable.

Commit:
Update inc_common.h

Fixed indentation.

Commit:
Update inc_common.h

Fixed indentation.

Commit:
Removed unnecessary parsing of salt2 in pure kernels

jsteube commented:
Please provide unit-test in tools/test_modules for this plugin. Shouldn't be too hard to realize there's plenty of other modules to copy the code from and then just adapt. But much more important is to remove the changes to inc_common.h. If you have a fast salt with esalt, there's also plenty of other modules than can be used as template.

Commit:
Removed unnecessary changes made to inc_common.h and made the proper adjustments in the OpenCL kernels to inflict the changes

davidrozen76 committed:
Merge branch 'master' into master

Commit:
Added a unit-test in tools/test_modules for hash-mode 3730

Commit:
Merge branch 'master' of https://github.com/davidrozen76/hashcat

Commit:
Updated hash-mode 3730 unit-test to work with two salts

davidrozen76 commented:
The unit-test fails on optimized kernels when salt length exceeds 16 characters.

Changes are to be made to support long salt lengths (a must for this hash-mode).

Any help will be appreciated :D

jsteube commented:
Any handling of data larger the block size of the underlaying primitive in question requires branching. This takes a lot of effort. Of course I can't merge an unfinished plugin. If you want me to merge this, please remove the optimized kernels, because the pure kernels should work. We can later add the optimized kernels.

Commit:
Removed hash-mode 3730 optimized kernel support

Optimized kernels requires handling on data larger than the block size of the underlaying primitive, thus more work is to be done to support optimized kernels.

jsteube commented:
Very cool, many thanks!
